### PR TITLE
Laboratory work #3, Dzyumenko Egor - 21FPL1

### PR DESCRIPTION
**Pull request by EgDzum**

### CODE REVIEW

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -195,7 +217,25 @@ def extract_pairs(tokens: tuple[int, ...], window_length: int) -> Optional[tuple
...
+    if not (tokens and isinstance(window_length, int) and window_length >= 2):
+        return None
+    window_tokens = tokens[:window_length]
+    pairs = []
+    start_slice = 0
+    for word in tokens:
+        for window_word in window_tokens:
+            if word == window_word:
+                continue
+            if word == window_word:
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:229]: why repeat these 2 lines?
```
if word == window_word:
    continue
```

**Review by marina-kaz — COMMENTED** [review]
It is a solid start, keep going!

**Review by marina-kaz — COMMENTED** [review]
nothing new to review :<

**Review by marina-kaz — COMMENTED** [review]
I hope at some point you'll decide to try for a greater mark

**Review by marina-kaz — COMMENTED** [review]
I remind that comments resolution is one of the criteria that neeeds to be fulfilled in order to be able to pass

File: lab_3_keywords_textrank/main.py [hunk 2]
```diff
@@ -564,7 +734,7 @@ def get_top_keywords(self, n_keywords: int) -> tuple[int, ...]:
             tuple[int, ...]
                 top n most important tokens in the encoded text
         """
-        pass
+        return tuple(sorted(self._scores, reverse=True, key=lambda key: self._scores[key])[:n_keywords])
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:737]: what if several vertices have the same score? how should they be sorted?

**Review by marina-kaz — COMMENTED** [review]
Great job!
A bit of improvement and you'll be all set

File: lab_3_keywords_textrank/main.py [hunk 3]
```diff
@@ -136,7 +151,12 @@ def encode(self, tokens: tuple[str, ...]) -> Optional[tuple[int, ...]]:
                 sequence of integer tokens
         In case of empty tokens input data, None is returned
         """
-        pass
+        if not tokens:
+            return None
+
+        self._learn_indices(tokens)
+        int_words = tuple((self._word2id[word] for word in tokens))
```
[the remark this answers is not in the document]
Comment (EgDzum) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -152,7 +172,12 @@ def decode(self, encoded_tokens: tuple[int, ...]) -> Optional[tuple[str, ...]]:
                 sequence of string tokens
         In case of out-of-dictionary input data, None is returned
         """
-        pass
+        for word in encoded_tokens:
+            if word not in self._id2word.keys():
+                return None
+
+        str_words = tuple((self._id2word[id_n] for id_n in encoded_tokens))
```
[the remark this answers is not in the document]
Comment (EgDzum) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py [hunk 5]
```diff
@@ -278,7 +350,16 @@ def calculate_inout_score(self, vertex: int) -> int:
                 number of incidental vertices
         If vertex is not present in the graph, -1 is returned
         """
-        pass
+        if vertex not in self._vertices:
+            return -1
+
+        index = self._vertices.index(vertex)
+        counting = 0
+        for row in self._matrix:
```
[the remark this answers is not in the document]
Comment (EgDzum) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py [hunk 6]
```diff
@@ -292,7 +373,15 @@ def fill_from_tokens(self, tokens: tuple[int, ...], window_length: int) -> None:
                 maximum distance between co-occurring tokens: tokens are considered co-occurring
                 if they appear in the same window of this length
         """
-        pass
+        pairs = extract_pairs(tokens, window_length)
+
+        for pair in pairs:
+            for word in pair:
```
[the remark this answers is not in the document]
Comment (EgDzum) [reply lab_3_keywords_textrank/main.py]: changed

File: lab_3_keywords_textrank/main.py [hunk 7]
```diff
@@ -385,7 +496,16 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
...
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+
+        for vertex in vertex1, vertex2:
+            if vertex not in self._edges.keys():
+                self._edges[vertex] = []
+
+        self._edges[vertex1].append(vertex2)
```
[the remark this answers is not in the document]
Comment (EgDzum) [reply lab_3_keywords_textrank/main.py:496]: changed

File: lab_3_keywords_textrank/main.py (hunk 2 above)
Comment (EgDzum) [reply lab_3_keywords_textrank/main.py]: changed

**Review by artyomtugaryov — APPROVED** [review]

**Pull request closed without merging**